### ISSUE TITLE
Can we not plug-in embeddings if synthesizing on a single-speaker model?

### ISSUE THREAD

**Issue by george-roussos** [mozilla/TTS#378]
As I probably mentioned on the discourse dev progress I made, I am interested in voice generation (new voice from other voices), so I thought I could extract the embeddings using the speaker encoder project and feed them to the TTS. This, however, seems to have some limitations:

- The embeddings extracted are sentence-specific, so I am guessing they are merely the spectograms of each sentence. So not actually speaker embeddings. I thought I could solve this by maybe getting the mean of all the embedding vectors extracted from one speaker and get the average in the latent space.

- I cannot seem to plug-in these embeddings in single-speaker models during inference time or training time, in order to manipulate the voice quality. In theory, what I want to do is see if I can get the mean of, say, 3 different speakers' embeddings and feed the mean output to the synthesizer and see if the voice changes in any way.

If I train a single-speaker model using the multi-speaker flag in the config.file, I get the .json file during training time, which only has the integers assigned to the speakers (in this case, just 0, since it is one speaker). But, there is no way to play with these embeddings.

So, my question is, I guess: how do I go about changing the voice of a speaker using the embeddings?

**Comment by erogol** [comment]
Embeddings you extracted from the encoder are sentence invariant and that is the essence of it. You can either provide embeddings per sentence or the mean vector at training time. Both should be fine.

In the inference time, you need to provide the right embedding vector to synthesize the corresponding voice. Or if you syn. voice for a new speaker you can generate its embedding vector with the encoder and provide it to the TTS model. In theory, TTS model should generalize well.

Using external embedding vectors is not supported yet in any of the released TTS versions. However, it should be easy to play with the code to do it.

- You need to load the right vector in TTSDataset.py
- Feed embed vec to the model in models/Tacotron.py (or Tacotron2.py)

And probably couple of more buzzers. 

Or another way to implement it is to initialize the speaker embedding layer with the encoder embeddings and freeze the layer. It should be easier.

**Comment by george-roussos** [comment]
Ah, okay, I thought it is not possible to provide with external embedding vectors, because there was no switch. Getting the mean of a speaker's voice embeddings seems trivial, then, however I think going the first way and modifying the code is probably only going to get me more headaches and I don't think hardcoding anything is a good idea either.  The second way sounds much more feasible, but I'd like some more pointers if you'd be so kind (this is my first time for many things for me).

**Comment by erogol** [comment]
you can provide a new speaker. you need to compute its embedding vector at inference time and provide that to the network.

**Comment by george-roussos** [comment]
Thanks, but how can we provide new speakers to an already-trained model? We have to pass the ID of the speaker in the `speaker_id` flag, and these IDs get assigned during training time. Or are you talking about editing `synthesis.py`? In that way, I would guess we would have to edit `id_to_torch`, and give it our embedding, but wouldn't that require hardcoding the embedding?

Otherwise, I took a look at the code and the suggestion about freezing the layer (thanks!) and I assume freezing is done in `TTS/models/tacotron2.py`, right? Because that is the only place I saw that the speaker embedding layer was initialized (line 42 — line 45). My question is whether it is initialized there, or at the end of the code.

If my first assumption (editing `synthesis.py`) is correct, I would be perfectly happy to try and make a multi-speaker version of the file in a pull request. 😀The second one, even.

**Comment by erogol** [comment]
At the inference time ignore the embedding layer and provide the computed embedding vector manually.

**Comment by george-roussos** [comment]
The implementation I am trying to edit is the one in the dev branch. I want to try training with these embeddings and not only infer. I can only see weights initialised in `TTS/models/tacotron2.py`.

**Comment by erogol** [comment]
I understand what you like to do. My suggestion was for only inference with a new speaker.

**Comment by george-roussos** [comment]
That one would also be great if it worked. Train a big multi-speaker model and then be able to adapt to a new speaker only using the external embeddings of them, extracted from Speaker Encoder.

**Comment by erogol** [comment]
It is right. initialization is done in Tacotron2.py.  

https://github.com/mozilla/TTS/blob/master/models/tacotron2.py#L43

**Comment by george-roussos** [comment]
Really sorry to be a bugger, but I cannot figure out for the life of me how to only initialize using my embeddings. Tried a lot of things but nothing seems to work. Any help from anyone would be greatly appreciated.

**Comment by Edresson** [comment]
@george-roussos  I'm working with multi-speaker on my branch I did the load embeddings for the nn.embedding layer (see here: https://github.com/Edresson/TTS/blob/dev-embedding/models/ tacotron.py), currently only supported tacotron.
To load your speaker embeddings you must pass the path of speakers.json in config.json (parameter speaker_embedding_file). 

The expected format for the speakers.json file is:
{name: {'id': id, 'embedding': embedding_list}, {...}, ...}

An example:
{"speaker_name_x": {'id': 0, 'embedding': [1,2,3,4,5,5,65,76,7,78,9]},
"speaker_name_y": {'id': 1, 'embedding': [5,2,3,4,5,5,65,76,7,78,9]}}

**Comment by george-roussos** [comment]
> @george-roussos I'm working with multi-speaker on my branch I did the load embeddings for the nn.embedding layer (see here: https://github.com/Edresson/TTS/blob/dev-embedding/models/ tacotron.py), currently only supported tacotron.
> To load your speaker embeddings you must pass the path of speakers.json in config.json (parameter speaker_embedding_file).
> 
> The expected format for the speakers.json file is:
> {name: {'id': id, 'embedding': embedding_list}, {...}, ...}
> 
> An example:
> {"speaker_name_x": {'id': 0, 'embedding': [1,2,3,4,5,5,65,76,7,78,9]},
> "speaker_name_y": {'id': 1, 'embedding': [5,2,3,4,5,5,65,76,7,78,9]}}

Amazing, thanks for the heads up! I managed to get them to load by modifying the model in the model dir directly, by initializing the speaker embedding layer for any amount of speakers and then freezing it. I did by creating a layer from pretrained weights, which was a torch float tensor I passed (dim 512). Since it freezes it is used for every speaker. The good thing about that approach is that it works both in Taco and Taco2 models. I do have to use a json file of speakers to make the synthesis use the embeddings.

**Comment by george-roussos** [comment]
Discussion is now developed [here](https://discourse.mozilla.org/t/multispeaker-development-progress/55342/16)
